### ISSUE TITLE
Regression in Pyright 1.1.373: valid union element disallowed

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#8519]
In Pyright 1.1.373, code with no type issues has started being disallowed. It seems like Pyright is matching against one of the elements in a union and eagerly disallowing it instead of checking the remaining union members. 

Issue was found in some of our tests using `pytest.mark.parametrize` (pytest 8.3.1). The tests are structured like this:

```python
import pytest

list1 = [1, 2, 3]
list2 = [1.0, 2.0, 3.0]
list3 = [True, False, True]
list4 = ["a", "b", "c"]

@pytest.mark.parametrize("arg1, arg2", list(zip(list1, list2)))  # No error
def test1(arg1, arg2):
    ... 

@pytest.mark.parametrize("arg1, arg2, arg3", list(zip(list1, list2, list3)))  # Error - Pyright complains about each argument to zip()
def test2(arg1, arg2, arg3):
    ... 

@pytest.mark.parametrize("arg1, arg2, arg3, arg4", list(zip(list1, list2, list3, list4)))  # No error
def test3(arg1, arg2, arg3, arg4):
    ... 
```

As far as I can tell, [the types in `pytest` for the `parametrize` decorator are reasonable](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/mark/structures.py#L462). It looks to me that when using 3 arguments, Pyright eagerly assumes the second arg to `parametrize` is `Iterable[ParameterSet]` instead of `Iterable[object]`, probably [because `ParameterSet` is a `NamedTuple` with 3 members](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/mark/structures.py#L75).  

There also seems to be a related bug **which fails in both versions** where if the second argument is `zip(list1, list2, list3)` instead of `list(zip(list1, list2, list3))` it is disallowed, despite the second argument to `parametrize` being `Iterable[ParameterSet | Sequence[object] | object]` (changing it to `Iterable[object]` makes Pyright accept the code). 

**Code or Screenshots**
I've trimmed out the relevant parts of `pytest` and copied them into a playground to try and create a minimal example for you.

* [This is the code passing in 372](https://pyright-play.net/?pyrightVersion=1.1.372&pythonVersion=3.12&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByAhhAKYAmAKgK4IA2pANFAMql3ACwAUKJFAGMwdBgJhIwKAM4A6IgCMBmbLnxsAjtVIoBTKAGFho8ZOYBJGKRAKGPHgLpEpUqAAUi1spZBsYACmIyKloGAEoALh4oaKgANyI6LSlw1lJNbV0AbTB5ACtSMQBdAGIoAB9CMBhfSkRSQqiYiA8Aa2SDIwKTFGLMgFlWgBEC3CIYXHKoAZAWhu4YzHIUqRgQSYJJUjtue0dnKAB9d09SVaQAL1JpluGha3GQcOK-a9vRh4jG6PJSYEODgQJOgHA5%2BL4LKBSdjARjghYeNAoEikdorNYVDRaHSkTJowqw%2BYQ6II%2BKJFEpCxWGw447I7y%2BSaYjI4nL5IqTVldfHg0JQAC0AD5UhwUjIxaVBVMhiN7rhRWLtjwfn9gGAwH4EB46SALqQUkctV4dZdXjKxrhmAB3FJ0JArTKoGD4qAADxtdpgmXkarozrg7vtwDoYDGzvOAc9eM%2BhOipUoAAsUaQoECoJbcC1MCgoABmADsACYoPJqPgYImoBAkORyAwoMAiEg6C5ULm8znwZqTmdLn4AER95i2lZ%2Bc5IBB%2BS3MF2hWedw2nY2kfuDqDD-xjidT13MOCz0Lz7tLldDj2j8eT6e75jnffbBZxitVmt1htNlvZ-NFgBkbZzzBLfAkHwKR4zAag6HIFAAHIYBkKAAANNwQzAXCIbMEMpax5AYbI8i5BDD21XUTygTdL1dOcYygLtiN7AcbwvbcXV3KiFloo0SIYsimKvKA4BvKigA)
* [This is the code failing in 373](https://pyright-play.net/?pyrightVersion=1.1.373&pythonVersion=3.12&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByAhhAKYAmAKgK4IA2pANFAMql3ACwAUKJFAGMwdBgJhIwKAM4A6IgCMBmbLnxsAjtVIoBTKAGFho8ZOYBJGKRAKGPHgLpEpUqAAUi1spZBsYACmIyKloGAEoALh4oaKgANyI6LSlw1lJNbV0AbTB5ACtSMQBdAGIoAB9CMBhfSkRSQqiYiA8Aa2SDIwKTFGLMgFlWgBEC3CIYXHKoAZAWhu4YzHIUqRgQSYJJUjtue0dnKAB9d09SVaQAL1JpluGha3GQcOK-a9vRh4jG6PJSYEODgQJOgHA5%2BL4LKBSdjARjghYeNAoEikdorNYVDRaHSkTJowqw%2BYQ6II%2BKJFEpCxWGw447I7y%2BSaYjI4nL5IqTVldfHg0JQAC0AD5UhwUjIxaVBVMhiN7rhRWLtjwfn9gGAwH4EB46SALqQUkctV4dZdXjKxrhmAB3FJ0JArTKoGD4qAADxtdpgmXkarozrg7vtwDoYDGzvOAc9eM%2BhOipUoAAsUaQoECoJbcC1MCgoABmADsACYoPJqPgYImoBAkORyAwoMAiEg6C5ULm8znwZqTmdLn4AER95i2lZ%2Bc5IBB%2BS3MF2hWedw2nY2kfuDqDD-xjidT13MOCz0Lz7tLldDj2j8eT6e75jnffbBZxitVmt1htNlvZ-NFgBkbZzzBLfAkHwKR4zAag6HIFAAHIYBkKAAANNwQzAXCIbMEMpax5AYbI8i5BDD21XUTygTdL1dOcYygLtiN7AcbwvbcXV3KiFloo0SIYsimKvKA4BvKigA)

**VS Code extension or command-line**
Found using commandline version of Pyright in CI when trying to upgrade from 1.1.371 to 1.1.373.

**Comment by erictraut** [comment]
Thanks for reporting. I've isolated the change in behavior to [this commit](https://github.com/microsoft/pyright/commit/74340854255fdf65e41a9ce568b64adb74981168).

Here's a simplified repro.

```python
from typing import Iterable, Sequence

list1 = [1]

class NT(tuple[list]):
    ...

x1: Iterable[NT | Sequence] = list(zip(list1))
x2: Iterable[NT | Sequence] = list(zip(list1, list1))

x3: Iterable[NT | Sequence] = [zip(list1, list1)]
```

I'll continue to investigate, but I wanted to record my findings so far (partly for my own benefit).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
